documentation/design-docs/ipc-protocol.md

```diff
@@ -0,0 +1,208 @@
+# IPC Protocol
+
+## Overview
+
+This spec describes a proposed IPC Protocol to be used for communicating with the runtime's diagnostics server from an external client over a platform-specific transport, e.g., Unix Domain Sockets. This protocol should be lightweight and generic, so as to be extensible for future use cases.  The primary use today is for EventPipe.
+
+## TODOs
+* formalize struct names
+* formalize error codes (Use HRESULT values?)
+
+## Inter-Process Communication Specification
+
+### Current Implementation (OLD)
+
+Single-purpose IPC potocol used exclusively for EventPipe functionality.  "Packets" in the current implementation are simply the `netperf` payloads and command/control is handled via `uint32` enum values sent one way with hardcoded responses expected.
+
+```c++
+enum class DiagnosticMessageType : uint32_t
+{
+    // EventPipe
+    StartEventPipeTracing = 1024, // To file
+    StopEventPipeTracing,
+    CollectEventPipeTracing, // To IPC
+};
+
+struct MessageHeader
+{
+    DiagnosticMessageType RequestType;
+    uint32_t Pid;
+};
+```
+
+```
+runtime <- client : MessageHeader { CollectEventPipeTracing }
+    error? -> no response
+runtime -> client : session ID 
+runtime -> client : event stream
+
+...
+
+runtime <- client : stop command
+runtime -> client : session id and stops
+```
+-----
+
+## Proposed Implementation (NEW)
+
+The IPC protocol will become a multi-purpose protocol to be used for communication to and from command servers running in the dotnet core runtime.  EventPipe is a good example of a service that will operate over this protocol.
+
+Connection (general):
+
+![](ipc-protocol-genericflow.svg)
+
+```
+runtime <- client : [ header ][ optional payload ]
+runtime -> client : [ header ][ optional payload ]
+optional
+    runtime <- client : [ optional stream of command specific data ]
+    ------or------
+    runtime -> client : [ optional stream of command specific data ]
+```
+
+EventPipe flow:
+```
+runtime <- client : [ ver; size; start_stream_command ][ stream config struct ]
+runtime -> client : [ ver; size; stream_started_response ][ stream started struct ]
+runtime -> client : [ stream of netperf data ]
+```
+
+### Transport considerations
+
+* All data will be sent little-endian
+* All commands will be sent in a packet with the defined header
+* Command specific data can be sent after a command ACK
+
+#### Transport
+
+The protocol will be communicated over a platform-specific transport.  On Unix/Linux based platforms, a Unix Domain Socket will be used, and on Windows, a Named Pipe will be used. 
```

Comment:
This section will need to add precise naming conventions to determine the OS primitive starting from a PID.